### PR TITLE
feat(cli): add support for jsxImportSourceTypes

### LINKED ISSUE #16653: NPM types: TSX warnings about JSX.IntrinsicElements in React
I have a TSX project with the following `deno.json` configuration:

```json
{
  "compilerOptions": {
    "jsx": "react-jsx",
    "jsxImportSource": "npm:react"
  }
}
```

In VS Code I see the following warning:

imaxe

This doesn't happen if the JSX import source is preact (`npm:preact`).

I think the reason is because the preact NPM package includes the types (https://github.com/preactjs/preact/blob/master/src/index.d.ts) so Deno can load them.

React needs to fetch the types from `@types/react` and this is not implemented in Deno. In the NPM roadmap (https://github.com/denoland/deno/issues/15960) I can see the uncompleted task `automatic acquisition of @types package for type checking`. 

Is there any plans to implement this at anytime soon? I think it should be quite useful. At this moment, the only way to use React in Deno without TSX errors is by importing from esm.sh.

### LINKED ISSUE #18203: deno check error resolving package config for npm react jsx-runtime
While importing the npm:@types packages as per #16653 silences errors from the language server when running deno check I see the following error:
```
% deno check index.tsx
Check file:///Users/lrowe/scratch/react-server-deno/index.tsx
error: Uncaught Error: Error resolving package config for 'npm:react@18.2.0/jsx-runtime': [ERR_PACKAGE_PATH_NOT_EXPORTED] Package subpath './jsx-runtime' is not defined by "exports" in '/Users/lrowe/Library/Caches/deno/npm/registry.npmjs.org/react/18.2.0/package.json' imported from '/Users/lrowe/Library/Caches/deno/npm/registry.npmjs.org/react/18.2.0/'
    at <anonymous> (ext:deno_tsc/99_main_compiler.js:592:28)
```

However in react's package.json an entry for './jsx-runtime' exists:
```json
  "exports": {
    ".": {
      "react-server": "./react.shared-subset.js",
      "default": "./index.js"
    },
    "./package.json": "./package.json",
    "./jsx-runtime": "./jsx-runtime.js",
    "./jsx-dev-runtime": "./jsx-dev-runtime.js"
  },
```

Files to reproduce:

index.tsx:
```ts
// See: https://github.com/denoland/deno/issues/16653
import type {} from "npm:@types/react@18";
import type {} from "npm:@types/react-dom@18";

import React from "npm:react@18";
import { renderToString } from "npm:react-dom@18/server";

function Hello() {
  return <div>Hello world!</div>;
}

console.log(renderToString(<Hello />));
```

deno.json:
```json
{
  "compilerOptions": {
    "jsx": "react-jsx",
    "jsxImportSource": "npm:react@18"
  }
}
```

```
% deno --version      
deno 1.31.2 (release, aarch64-apple-darwin)
v8 11.0.226.19
typescript 4.9.4
```

### LINKED ISSUE #20455: npm: automatic acquisition of `@types` package for type checking
This was already on the roadmap but with no issue.

You can usually work around this with `// @deno-types="npm:@types/<package>"` but that's not always an option. E.g. for `"jsxImportSource": "npm:react"` in your compiler options, there's no import to add the directive to. So it's a big obstacle for react-jsx users.

### PR DESCRIPTION
**Pull request by lucacasonato**
TODO:

- ~~understand why LSP has special handling of `jsxImportSource` in compiler
  options as compared to `deno check`~~
- ~~ensure that @jsxImportSourceTypes pragma works in LSP~~ ✅ 
- ~~ensure that @jsxImportSourceTypes pragma works in `deno check`~~ ✅ 
- ~~tests~~

Closes #16653
Closes #18203
Closes #20455

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -96,14 +96,23 @@ cache_control = "=0.2.0"
...
 dlopen2 = "0.6.1"
 ecb = "=0.1.2"
-elliptic-curve = { version = "0.13.4", features = ["alloc", "arithmetic", "ecdh", "std", "pem"] }
+elliptic-curve = { version = "0.13.4", features = [
+  "alloc",
+  "arithmetic",
+  "ecdh",
+  "std",
+  "pem",
+] }
```
Comment (dsherret) [inline Cargo.toml:115]: This kind of formatting takes up a lot of vertical space for not very useful information.

File: cli/emit.rs [hunk 2]
```diff
@@ -100,17 +100,9 @@ impl Emitter {
...
-        Err(parsed_source) => {
-          // transpile_owned is more efficient and should be preferred
-          debug_assert!(false, "Transpile owned failed.");
-          parsed_source
-            .transpile(&self.transpile_options, &self.emit_options)?
-        }
-      };
+      let transpiled_source = parsed_source
+        .transpile(&self.transpile_options, &self.emit_options)?
+        .into_source();
```
Comment (dsherret) [inline cli/emit.rs:105]: There is a variant on this return value that says whether it had to clone or whether it was owned. Can you revert the code more back to the original and use that new value? We want to ensure we're not cloning when emitting here.

Comment (lucacasonato) [reply cli/emit.rs]: Done in #23421

File: Cargo.toml [hunk 3]
```diff
@@ -43,7 +43,7 @@ license = "MIT"
 repository = "https://github.com/denoland/deno"
 
 [workspace.dependencies]
-deno_ast = { version = "=0.36.2", features = ["transpiling"] }
+deno_ast = { version = "0.37.0", features = ["transpiling"] }
```
Comment (dsherret) [inline Cargo.toml:46]: An example of how unpinning could break `cargo install deno` is the 0.37.1 release of deno_ast.

**Review by dsherret — APPROVED** [review]
LGTM

**Pull request merged**